### PR TITLE
Add HttpContextExtensions for GetActionContextAsync

### LINKED ISSUE #16554: move GetActionContextAsync and its related stuff to abstract place
### Is your feature request related to a problem? Please describe.
under public class SeoMetaSettingsHandler and its related 
 private readonly IActionContextAccessor _actionContextAccessor;
 private readonly IHttpContextAccessor _httpContextAccessor;
 private readonly IUrlHelperFactory _urlHelperFactory;

### Describe the solution you'd like
move 
 private readonly IActionContextAccessor _actionContextAccessor;
 private readonly IHttpContextAccessor _httpContextAccessor;
 private readonly IUrlHelperFactory _urlHelperFactory;
and
GetActionContextAsync
to a topService, because many modules  will use the same logic.

### PR DESCRIPTION
**Pull request by infofromca**
Fixes #16554

### CODE REVIEW

**Review by Piedone — CHANGES_REQUESTED** [review]
Please merge to adapt to file-scoped namespaces (https://github.com/OrchardCMS/OrchardCore/pull/16539), otherwise looks good except for the nitpick.

File: src/OrchardCore/OrchardCore.DisplayManagement/Extensions/HttpContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        var actionContext = httpContext.RequestServices.GetService<IActionContextAccessor>()?.ActionContext;
+
+        if (actionContext != null)
+        {
+            return actionContext;
+        }
+
+        return await GetActionContextAsync(httpContext);
+    }
+    public static async Task<ActionContext> GetActionContextAsync(this HttpContext httpContext)
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.DisplayManagement/Extensions/HttpContextExtensions.cs:25]:
```suggestion

    public static async Task<ActionContext> GetActionContextAsync(this HttpContext httpContext)
```

**Review by Piedone — APPROVED** [review]

File: src/OrchardCore/OrchardCore.DisplayManagement/Extensions/HttpContextExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        var httpContext = httpContextAccessor.HttpContext;
+        var actionContext = httpContext.RequestServices.GetService<IActionContextAccessor>()?.ActionContext;
+
+        if (actionContext != null)
+        {
+            return actionContext;
+        }
+
+        return await GetActionContextAsync(httpContext);
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.DisplayManagement/Extensions/HttpContextExtensions.cs:23]: remove `await` and `async` and just return a task instead.

File: src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorShapeTemplateViewEngine.cs [hunk 3]
```diff
@@ -101,7 +98,7 @@ private async Task<IHtmlContent> RenderRazorViewAsync(string viewName, DisplayCo
 
     public async Task<string> RenderViewToStringAsync(string viewName, object model, IViewEngine viewEngine)
     {
-        var actionContext = await GetActionContextAsync();
+            var actionContext = await _httpContextAccessor.GetActionContextAsync();
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorShapeTemplateViewEngine.cs:101]:
```suggestion
        var actionContext = await _httpContextAccessor.GetActionContextAsync();
```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]
Why do these two extensions need to be Async?

File: src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs [hunk 4]
```diff
@@ -192,28 +193,12 @@ public static async Task<ViewContext> GetViewContextAsync(LiquidTemplateContext
         if (actionContext == null)
         {
             var httpContext = context.Services.GetRequiredService<IHttpContextAccessor>().HttpContext;
-            actionContext = await GetActionContextAsync(httpContext);
+                actionContext = await httpContext.GetActionContextAsync();
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.DisplayManagement.Liquid/LiquidViewTemplate.cs:196]:
```suggestion
            actionContext = await httpContext.GetActionContextAsync();
```

File: src/OrchardCore.Modules/OrchardCore.Seo/Handlers/SeoMetaPartHandler.cs [hunk 5]
```diff
@@ -74,7 +71,7 @@ public override Task GetContentItemAspectAsync(ContentItemAspectContext context,
 
             var actionContext = _actionContextAccessor.ActionContext;
 
-            actionContext ??= await GetActionContextAsync(_httpContextAccessor.HttpContext);
+                actionContext ??= await _httpContextAccessor.GetActionContextAsync();
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Seo/Handlers/SeoMetaPartHandler.cs:74]:
```suggestion
            actionContext ??= await _httpContextAccessor.GetActionContextAsync();
```

File: src/OrchardCore/OrchardCore.DisplayManagement/Extensions/HttpContextExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        var filters = httpContext.RequestServices.GetServices<IAsyncViewActionFilter>();
+
+        foreach (var filter in filters)
+        {
+            filter.OnActionExecutionAsync(actionContext);
+        }
+
+        return Task.FromResult(actionContext);
+    }
+}
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.DisplayManagement/Extensions/HttpContextExtensions.cs:41]: The `OnActionExecutionAsync` call must be awaited 

 
```suggestion
    public static async Task<ActionContext> GetActionContextAsync(this HttpContext httpContext)
    {
        var routeData = new RouteData();
        routeData.Routers.Add(new RouteCollection());

        var actionContext = new ActionContext(httpContext, routeData, new ActionDescriptor());
        var filters = httpContext.RequestServices.GetServices<IAsyncViewActionFilter>();

        foreach (var filter in filters)
        {
            await filter.OnActionExecutionAsync(actionContext);
        }

        return actionContext;
    }
}
```

**Pull request merged**